internal/rfc7512/rfc7512.go

```diff
@@ -0,0 +1,196 @@
+// Package rfc7512 provides a parser for the PKCS #11 URI format as specified in
+// RFC 7512: The PKCS #11 URI Scheme. Additionally, it provides a wrapper around
+// the crypto11 package for loading a key pair as a crypto.Signer object.
+package rfc7512
+
+import (
+	"crypto"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/ThalesIgnite/crypto11"
+)
+
+// PKCS11URI contains the information for accessing a PKCS #11 object, such as
+// a public key, private key, or a certificate. The general form represented is:
+//
+//	pkcs11:path-attr[?query-attr]
+//
+// Path and query attributes are delimited by ';' and '&', respectively.
+// For more information read: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct {
+	// path attributes:
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS
+	ID     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotID    uint   //           slot-id <- CK_SLOT_ID
+
+	// query attributes:
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+
+	// Vendor specific query attributes:
+	MaxSessions int // max-sessions
+}
+
+// ParsePKCS11URI decodes a PKCS#11 URI as defined in RFC 7512 into an PKCS11URI object.
+// An error is returned if the input string does not appear to follow the rules
+// or if there are unrecognized path or query attributes.
+func ParsePKCS11URI(uri string) (*PKCS11URI, error) {
+	// Check that the URI matches the specification from RFC 7512:
+	aChar := "[a-z-_]"
+	pChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=&]"
+	pAttr := aChar + "+=" + pChar + "+"
+	pClause := "(" + pAttr + ";)*(" + pAttr + ")"
+	qChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=/\\?\\|]"
+	qAttr := aChar + "+=" + qChar + "+"
+	qClause := "(" + qAttr + "&)*(" + qAttr + ")"
+	r, _ := regexp.Compile("^pkcs11:" + pClause + "(\\?" + qClause + ")?$")
+
+	if !r.MatchString(uri) {
```

Comment:
I believe it's possible to parse and validate this format without regex. Something like `url.ParseQuery`. Using that as a building block might make things a lot simpler.

Reply:
I think that's a good idea, but there are edge cases that will make it just as complicated. For instance path attributes can contain "&".